src/tests.rs

```diff
@@ -0,0 +1,176 @@
+}
+
+#[test]
+fn fill_zero() {
+    getrandom(&mut []).unwrap();
+}
+#[test]
+fn fill_zero_uninit() {
+    getrandom_uninit(&mut []).unwrap();
+}
+#[test]
+fn fill_small() {
+    check_small(0, wrapped_fill);
+}
+#[test]
+fn fill_small_uninit() {
+    check_small(MaybeUninit::uninit(), getrandom_uninit);
+}
+#[test]
+fn fill_large() {
+    check_bits(1_000, 0, wrapped_fill);
+}
+#[test]
+fn fill_large_uninit() {
+    check_bits(1_000, MaybeUninit::uninit(), getrandom_uninit);
+}
+#[test]
+fn fill_huge() {
+    check_bits(1_000_000, 0, wrapped_fill);
+}
+#[test]
+fn fill_huge_uninit() {
+    check_bits(1_000_000, MaybeUninit::uninit(), getrandom_uninit);
+}
+
+// On WASM, the thread API always fails/panics
+#[cfg(not(any(target_arch = "wasm32", target_arch = "wasm64")))]
+#[test]
+fn multithreading() {
+    use std::{sync::mpsc, thread};
+
+    let mut txs = vec![];
+    for _ in 0..20 {
+        let (tx, rx) = mpsc::channel();
+        txs.push(tx);
+
+        thread::spawn(move || {
+            // wait until all the tasks are ready to go.
+            rx.recv().unwrap();
+            let mut v = [0u8; 1000];
+
+            for _ in 0..100 {
+                getrandom(&mut v).unwrap();
+                thread::yield_now();
+            }
+        });
+    }
+
+    // start all the tasks
+    for tx in txs.iter() {
+        tx.send(()).unwrap();
+    }
+}
+
+// A simple Linear Congruential Generator (LCG) that passes our RNG tests.
+const A: u32 = 134775813;
+const C: u32 = 1;
+static X: AtomicU32 = AtomicU32::new(0);
+
+fn gen_lcg() -> u32 {
+    X.fetch_update(Ordering::Relaxed, Ordering::Relaxed, |x| {
+        Some(x.wrapping_mul(A).wrapping_add(C))
+    })
+    .unwrap()
+}
+
+// Custom error returned when we get an input of length 1337
+const CUSTOM_CODE: u32 = Error::CUSTOM_START + 1337;
+
+#[allow(dead_code)]
+fn insecure_rng(buf: &mut [u8]) -> Result<(), Error> {
+    // We guarantee no implementation is called with an empty buffer.
+    assert!(!buf.is_empty());
+    if buf.len() == 1337 {
+        return Err(NonZeroU32::new(CUSTOM_CODE).unwrap().into());
+    }
+
+    let mut chunks = buf.chunks_exact_mut(4);
+    for chunk in chunks.by_ref() {
+        chunk.copy_from_slice(&gen_lcg().to_le_bytes());
+    }
+    let tail = chunks.into_remainder();
+    tail.copy_from_slice(&gen_lcg().to_le_bytes()[..tail.len()]);
+    Ok(())
+}
+
+// Test registering a custom implementation, even on supported platforms.
+#[cfg(feature = "custom")]
+register_custom_getrandom!(insecure_rng);
+
+// On a supported platform, make sure the custom implementation isn't used.
+#[cfg(all(feature = "custom", target_os = "linux"))]
+#[test]
+fn custom_not_used() {
+    getrandom(&mut [0; 1337]).unwrap();
+}
+
+// On an unsupported platform, make sure the custom implementation is used.
+#[cfg(all(feature = "custom", target_arch = "wasm32", target_os = "unknown"))]
+#[test]
+fn custom_used() {
+    let err = getrandom(&mut [0; 1337]).unwrap_err();
```

Comment:
In this kind of test, we should just use a mock custom implementation that has state that we can directly query (e.g. using Mutex-protected global state), instead of indirectly interpreting its output.

Reply:
I moved the custom RNG changes to #473 and simplified the implementation there. It doesn't use a mock (I couldn't figure out a clean way to do that in a test).